layers/ipfix.go

```diff
@@ -0,0 +1,773 @@
+/* LICENSE */
+/*
+This layer decodes IPFIX version 10 datagrams.
+
+The specification can be found here: https://datatracker.ietf.org/doc/html/rfc7011
+
+This decoder only decode Data Records if the datagram also contains Template or
+Option Template associated to it. Otherwise, Data Records will only contains
+bytes.
+This decoder does no parse well known Data fields defined in IANA-IPFIX
+https://www.iana.org/assignments/ipfix/ipfix.xhtml
+
+TODO:
+- provide a method to decode Data Record with raw bytes and appropriate Template
+  or Option Template
+- suppoer encoding method `SerializeTo`
+- use gopacket DecodeFeedback
+- benchmark a more consequent pcap file
+*/
+
+package layers
+
+import (
+	"encoding/binary"
+	"fmt"
+
+	"github.com/google/gopacket"
+)
+
+const (
+	// Version word in the Message Header
+	Version uint16 = 0x000a
+	// EnterpriseBit used in the Field Specifier
+	EnterpriseBit uint16 = 0x8000
+	// VariableLength used in the Field Specifier
+	VariableLength uint16 = 0xffff
+)
+
+// Register IPFIX layer
+
+func decodeIPFIX(data []byte, p gopacket.PacketBuilder) error {
+	ipfix := &IPFIX{
+		Templates: make(map[IPFIXSetIDType]IPFIXTemplateAccessor),
+	}
+
+	err := ipfix.DecodeFromBytes(data, p)
+	if err != nil {
+		return err
+	}
+	p.AddLayer(ipfix)
+	p.SetApplicationLayer(ipfix)
+
+	return nil
+}
+
+// IPFIXFieldFunc function call to iter Template/OptionTemplate Field Specifiers
+type IPFIXFieldFunc func(fs *IPFIXFieldSpecifier) error
+
+// IPFIXTemplateAccessor interface to access Template and OptionTemplate
+type IPFIXTemplateAccessor interface {
+	GetID() IPFIXSetIDType
+	GetMinDataRecordLength() uint16
+	IterFieldSpecifier(fn IPFIXFieldFunc) error
+}
+
+// IPFIX is the outermost container which holds packet header and holds
+// at least one template or one data set
```

Comment:
A figure that depicts IPFIX packet which can hold multiple template/data sets would be good.